ICE: extracting field of non-tuple non-adt

This is a test case reduction of one of the Tock errors.

When I try to verify the code:
```Rust
enum Color {
    Red,
    Green,
    Blue,
}

fn add_colorcode(num: Result<usize, ()>, color: &Color) -> Result<usize, ()> {
    num.map(|n| match color {
        Color::Red => n + 1,
        Color::Green => n + 2,
        Color::Blue => n + 3,
    })
}
```

It fails with:
```
error[E0999]: internal flux error: crates/flux-middle/src/lib.rs:576:22
   --> src/main.rs:621:1
    |
621 | fn add_colorcode(num: Result<usize, ()>, color: &Color) -> Result<usize, ()> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: extracting field of non-tuple non-adt: PlaceTy { ty: Closure add_colorcode::{closure#0}<i32, fn((usize,)) -> usize, (&'?11 &'?12 Color,)>, variant_index: None }
```

In general, this seems to happen anytime there is a cast match inside a closure.

--- a/crates/flux-infer/src/refine_tree.rs
+++ b/crates/flux-infer/src/refine_tree.rs
@@ -757,19 +757,17 @@ mod pretty {
         cx: &PrettyCx,
         f: &mut fmt::Formatter<'_>,
     ) -> fmt::Result {
-        let mut f = with_padding(f);
         define_scoped!(cx, f);
         match children {
             [] => w!(" true"),
             [n] => {
                 if n.borrow().is_head() {
-                    w!(" ")?;
+                    w!(" {:?}", n)
                 } else {
-                    w!("\n")?;
+                    w!(with_padding(f), "\n{:?}", n)
                 }
-                w!("{:?}", n)
             }
-            _ => w!("\n{:?}", join!("\n", children)),
+            _ => w!(with_padding(f), "\n{:?}", join!("\n", children)),
         }
     }
 

--- a/crates/flux-middle/src/lib.rs
+++ b/crates/flux-middle/src/lib.rs
@@ -48,7 +48,7 @@ use flux_macros::fluent_messages;
 pub use flux_rustc_bridge::def_id_to_string;
 use flux_rustc_bridge::{
     mir::{LocalDecls, PlaceElem},
-    ty,
+    ty::{self, GenericArgsExt},
 };
 use flux_syntax::surface::{self, NodeId};
 use global_env::GlobalEnv;
@@ -573,7 +573,8 @@ impl PlaceTy {
                 Ok(ty.subst(args))
             }
             ty::TyKind::Tuple(tys) => Ok(tys[f.index()].clone()),
-            _ => Err(query_bug!("extracting field of non-tuple non-adt: {self:?}")),
+            ty::TyKind::Closure(_, args) => Ok(args.as_closure().upvar_tys()[f.index()].clone()),
+            _ => Err(query_bug!("extracting field of non-tuple non-adt non-closure: {self:?}")),
         }
     }
 }

--- a/crates/flux-middle/src/queries.rs
+++ b/crates/flux-middle/src/queries.rs
@@ -756,6 +756,7 @@ impl<'a> Diagnostic<'a> for QueryErr {
 }
 
 impl<'a> Diagnostic<'a> for QueryErrAt {
+    #[track_caller]
     fn into_diag(
         self,
         dcx: rustc_errors::DiagCtxtHandle<'a>,
@@ -763,42 +764,47 @@ impl<'a> Diagnostic<'a> for QueryErrAt {
     ) -> rustc_errors::Diag<'a, ErrorGuaranteed> {
         use crate::fluent_generated as fluent;
 
-        rustc_middle::ty::tls::with(|tcx| {
-            let mut diag = match self.err {
-                QueryErr::Unsupported { def_id, err, .. } => {
-                    let mut diag =
-                        dcx.struct_span_err(self.span, fluent::middle_query_unsupported_at);
-                    diag.arg("kind", tcx.def_kind(def_id).descr(def_id));
-                    if let Some(def_ident_span) = tcx.def_ident_span(def_id) {
-                        diag.span_note(def_ident_span, fluent::_subdiag::note);
+        rustc_middle::ty::tls::with_opt(
+            #[track_caller]
+            |tcx| {
+                let tcx = tcx.expect("no TyCtxt stored in tls");
+                let mut diag = match self.err {
+                    QueryErr::Unsupported { def_id, err, .. } => {
+                        let mut diag =
+                            dcx.struct_span_err(self.span, fluent::middle_query_unsupported_at);
+                        diag.arg("kind", tcx.def_kind(def_id).descr(def_id));
+                        if let Some(def_ident_span) = tcx.def_ident_span(def_id) {
+                            diag.span_note(def_ident_span, fluent::_subdiag::note);
+                        }
+                        diag.note(err.descr);
+                        diag
                     }
-                    diag.note(err.descr);
-                    diag
-                }
-                QueryErr::Ignored { def_id } => {
-                    let mut diag = dcx.struct_span_err(self.span, fluent::middle_query_ignored_at);
-                    diag.arg("kind", tcx.def_kind(def_id).descr(def_id));
-                    diag.arg("name", def_id_to_string(def_id));
-                    diag.span_label(self.span, fluent::_subdiag::label);
-                    diag
-                }
-                QueryErr::InvalidAssocReft { .. } => {
-                    let mut diag =
-                        dcx.struct_span_err(self.span, fluent::middle_query_invalid_assoc_reft_at);
-                    diag.code(E0999);
-                    diag
-                }
-                QueryErr::InvalidGenericArg { .. }
-                | QueryErr::Emitted(_)
-                | QueryErr::Bug { .. } => {
-                    let mut diag = self.err.into_diag(dcx, level);
-                    diag.span(self.span);
-                    diag
-                }
-            };
-            diag.code(E0999);
-            diag
-        })
+                    QueryErr::Ignored { def_id } => {
+                        let mut diag =
+                            dcx.struct_span_err(self.span, fluent::middle_query_ignored_at);
+                        diag.arg("kind", tcx.def_kind(def_id).descr(def_id));
+                        diag.arg("name", def_id_to_string(def_id));
+                        diag.span_label(self.span, fluent::_subdiag::label);
+                        diag
+                    }
+                    QueryErr::InvalidAssocReft { .. } => {
+                        let mut diag = dcx
+                            .struct_span_err(self.span, fluent::middle_query_invalid_assoc_reft_at);
+                        diag.code(E0999);
+                        diag
+                    }
+                    QueryErr::InvalidGenericArg { .. }
+                    | QueryErr::Emitted(_)
+                    | QueryErr::Bug { .. } => {
+                        let mut diag = self.err.into_diag(dcx, level);
+                        diag.span(self.span);
+                        diag
+                    }
+                };
+                diag.code(E0999);
+                diag
+            },
+        )
     }
 }
 

--- a/crates/flux-refineck/src/checker.rs
+++ b/crates/flux-refineck/src/checker.rs
@@ -1778,6 +1778,7 @@ pub(crate) mod errors {
     }
 
     impl<'a> Diagnostic<'a> for CheckerError {
+        #[track_caller]
         fn into_diag(
             self,
             dcx: rustc_errors::DiagCtxtHandle<'a>,

--- a/tests/tests/pos/surface/issue-809.rs
+++ b/tests/tests/pos/surface/issue-809.rs
@@ -1,0 +1,22 @@
+use flux_rs::*;
+
+#[refined_by(val: int)]
+enum Color {
+    #[variant(Color[1])]
+    Red,
+    #[variant(Color[2])]
+    Green,
+    #[variant(Color[3])]
+    Blue,
+}
+
+#[sig(fn(_, &Color[@c]) -> Result<usize{v: v >= c}, ()>)]
+fn add_colorcode(num: Result<usize, ()>, color: &Color) -> Result<usize, ()> {
+    num.map(|n| {
+        match color {
+            Color::Red => n + 1,
+            Color::Green => n + 2,
+            Color::Blue => n + 3,
+        }
+    })
+}

